### ISSUE TITLE
getAllData not working on Win32 / arm64

### ISSUE THREAD

**Issue by GrumpyCrouton** [sebhildebrandt/systeminformation#407]
**Describe the bug**

When I compile my app using platform Win32 and arch arm64, getAllData() does nothing.

**To Reproduce**
1. compile using `npx electron-rebuild . --arch=arm64 && npx electron-packager . --platform=win32 --arch=arm64 --overwrite`

```
$( document ).ready(function() {
	$('#dc_version').text(' - Version ' + dc_version);
	
	si.getAllData(function(systemdata) {
		console.log('got system data');
		Object.assign(pulled_systemdata, systemdata);
	});
});
```

3. My `$('#dc_version')` code is running and working, but I never get the `got system data` console log.

**Environment (please complete the following information):**
 - systeminformation package version: 4.22.3
 - OS: Windows 10
 - Hardware: Lenovo Flex 5G

--

If I compile as `x64` or `ia32`, it works on this same device.

**Comment by sebhildebrandt** [comment]
@GrumpyCrouton can you provide a link to the repository with full code ... 

My feeling is, that you are calling `is.getAllData()` in the wrong place ... (render process and not main process ...)

**Comment by GrumpyCrouton** [comment]
@sebhildebrandt I can't really provide the full code, but will explain where I run this code.

In my `main.js` file, I run some code to load my `.html` file (`win.loadFile('app/master.html')`) which contains my first block of code.

The code works if I compile using `x64` or `ia32`. I've been using systeminformation like this for a few years now.

**Comment by sebhildebrandt** [comment]
@GrumpyCrouton  OK, seems that this is due to the arm platform ... in this case we need to find out which function blocks the output ... can you first try to run `si.getDynamicData()` and `si.getStaticData()` separately? 

if this is related to si.getStaticData() then can you run all these separately:

```
si.system()
si.bios()
si.baseboard()
si.chassis()
si.osInfo()
si.uuid()
si.versions()
si.cpu()
si.cpuFlags()
si.graphics()
si.networkInterfaces()
si.memLayout()
si.diskLayout()
```

And the the same for the dynamic functions:

```
si.cpuCurrentspeed()
si.users()
si.currentLoad()
si.networkStats(iface)
si.networkConnections()
si.mem()
si.battery()
si.fsSize()
si.inetLatency()
```

thank you in advance

**Comment by GrumpyCrouton** [comment]
Neither of the 2 functions you mentioned seem to do anything.

I tested each method separately using this code:

```
console.log('running');
si.system(function(systemdata) {
	console.log('system', systemdata);
});
si.bios(function(systemdata) {
	console.log('bios', systemdata);
});
si.baseboard(function(systemdata) {
	console.log('baseboard', systemdata);
});
si.chassis(function(systemdata) {
	console.log('chassis', systemdata);
});
si.osInfo(function(systemdata) {
	console.log('osInfo', systemdata);
});
si.uuid(function(systemdata) {
	console.log('uuid', systemdata);
});
si.versions(function(systemdata) {
	console.log('versions', systemdata);
});
si.cpu(function(systemdata) {
	console.log('cpu', systemdata);
});
si.cpuFlags(function(systemdata) {
	console.log('cpuFlags', systemdata);
});
si.graphics(function(systemdata) {
	console.log('graphics', systemdata);
});
si.networkInterfaces(function(systemdata) {
	console.log('networkInterfaces', systemdata);
});
si.memLayout(function(systemdata) {
	console.log('memLayout', systemdata);
});
si.diskLayout(function(systemdata) {
	console.log('diskLayout', systemdata);
});
si.cpuCurrentspeed(function(systemdata) {
	console.log('cpuCurrentspeed', systemdata);
});
si.users(function(systemdata) {
	console.log('users', systemdata);
});
si.currentLoad(function(systemdata) {
	console.log('currentLoad', systemdata);
});
si.networkStats(function(systemdata) {
	console.log('networkStats', systemdata);
});
si.networkConnections(function(systemdata) {
	console.log('networkConnections', systemdata);
});
si.mem(function(systemdata) {
	console.log('mem', systemdata);
});
si.battery(function(systemdata) {
	console.log('battery', systemdata);
});
si.fsSize(function(systemdata) {
	console.log('fsSize', systemdata);
});
si.inetLatency(function(systemdata) {
	console.log('inetLatency', systemdata);
});
```

and I get nothing logged in my console except for `running`.

**Comment by sebhildebrandt** [comment]
@GrumpyCrouton thank you for testing! So I need to make some research on windows on arm ;-)

**Comment by GrumpyCrouton** [comment]
@sebhildebrandt It's strange because if I compile as `x86` or `ia32`, it works fine. I wish I could just keep using that arch, but unfortunately I can't due to some requirements of the project.

**Comment by sebhildebrandt** [comment]
@GrumpyCrouton very clear ... will have a look on it.

**Comment by Skillnter** [comment]
Running the following on window is not returning processes but it is returning while running on windows bash(ubuntu).
> si.getDynamicData("*","*")


**OS Info:**

> distro: "Microsoft Windows 10 Pro"
> kernel: "10.0.18363"
> release: "10.0.18363"
> arch: "x64"
> build: "18363"

**Comment by ibc** [comment]
This is happening to us in Electron 11 in Windows 10 (I don't think it's ARM at all). We use the following calls:

* `systemInformation.system()`
* `await systemInformation.cpu()`
* `await systemInformation.graphics()`

**Comment by ibc** [comment]
I've added some logs to see which one exactly never resolves, will update here when we have info.

**Comment by sebhildebrandt** [comment]
@ibc I can't see any logs or output ... can you please attach further information? Thank you so much.

**Comment by ibc** [comment]
I meant that we have added those logs into our app so we'll get them eventually when any client experiments it. Unfortunately I cannot reproduce the issue.

**Comment by ibc** [comment]
NOTE: Still waiting for those logs, do not close the issue, please :)

**Comment by sebhildebrandt** [comment]
@ibc shure! Will not close it. Need to get Windows ARM running on one of my raspberries to be able to test it here.

**Comment by ibc** [comment]
Actually the issue happened to me (not to me, but to a user of my app) running Windows in regular PC computer (no ARM at all).

**Comment by Skillnter** [comment]
> @ibc I can't see any logs or output ... can you please attach further information? Thank you so much.

Hi,

Can you tell me where to add logs? I'll be happy to share the logs if it can help. I am also stuck and might not be able to resolve by myself.

**Comment by ibc** [comment]
> Can you tell me where to add logs?

I don't understand what you mean, sorry.

**Comment by Skillnter** [comment]
> > Can you tell me where to add logs?
> 
> I don't understand what you mean, sorry.
`if (!_windows) {
        processes.processes().then(res => {
          data.processes = res;
          functionProcessed();
        });
      }`

I found this code that prevents processes to get on a windows machine. 

Just to test, I update it like the following:
`if (!_windows) {
        processes.processes().then(res => {
          data.processes = res;
          functionProcessed();
        });
      }else{console.log(processes)
        processes.processes().then(res => {
          data.processes = res;
          functionProcessed();
        });
      }`
And voila windows starts to get processes but I am not sure about the IF statement. What I have changed may have consequences which I am not able to see now but you had some reason for putting it. 

If you can guide me, I can be of better help and this package can grow better. Or I just can just tinkle around which can waste time as you have gone through some issues which I might face.

**Comment by Skillnter** [comment]
> > Can you tell me where to add logs?
> 
> I don't understand what you mean, sorry.

I am sorry, I quoted the wrong person, 





@sebhildebrandt  I found this code that prevents processes to get on a windows machine. 

`if (!_windows) {
        processes.processes().then(res => {
          data.processes = res;
          functionProcessed();
        });
      }`


Just to test, I update it like the following:
`if (!_windows) {
        processes.processes().then(res => {
          data.processes = res;
          functionProcessed();
        });
      }else{console.log(processes)
        processes.processes().then(res => {
          data.processes = res;
          functionProcessed();
        });
      }`
And voila windows starts to get processes but I am not sure about the IF statement. What I have changed may have consequences which I am not able to see now but you had some reason for putting it. 

If you can guide me, I can be of better help and this package can grow better. Or I just can just tinkle around which can waste time as you have gone through some issues which I might face.

**Comment by ibc** [comment]
Sorry, I think the issue is clearly explained in comments above. This is not about looking for a specific workaround for your specific code or use case. This is about providing proper reports that help fixing the real issue.

**Comment by Skillnter** [comment]
> Sorry, I think the issue is clearly explained in comments above. This is not about looking for a specific workaround for your specific code or use case. This is about providing proper reports that help fixing the real issue.

Yes, and to get a proper fix, we need to understand the issue which can be hard as a few of the things are hard to understand. I just gave an example of which I find weird. It was not a workaround.

**Comment by Skillnter** [comment]
> Sorry, I think the issue is clearly explained in comments above. This is not about looking for a specific workaround for your specific code or use case. This is about providing proper reports that help fixing the real issue.

Yes, and to get a proper fix, we need to understand the issue which can be hard as a few of the things are hard to understand. I just gave an example of which I find weird. It was not a workaround. I am also trying to help. 

If you want me to stop, I'll open a new issue. No hard feelings.

**Comment by ibc** [comment]
I'm not project maintainer.

**Comment by Skillnter** [comment]
> I'm not project maintainer.

That's why I tagged him if you check the comment.

**Comment by sebhildebrandt** [comment]
@Skillnter @ibc ... having a look on your comments (@Skillnter) I will first try to go into the code to see what could be the issue ... I thought that all of this is related to Windows running on ARM (which - In your case - is not true, right). 

Your specific issue is (if I am right)
- systeminformation on regular windows non ARM (but 32bit??)
- `si.getDynamicData("","", cb)` and `si.getAllData("","",cb)` not returning processes

So I will check first here on my side, why there is a reason not getting processes. Will come back end of next week, as I have - due to nome office regulation - currently very limited access to my windows machines.

**Comment by ibc** [comment]
> * systeminformation on regular windows non ARM (but 32bit??)

I don't know yet. If any user of my app experiences this issue I'll get a complete report with logs and system information. (Un)fortunately it doesn't happen yet :(

**Comment by Skillnter** [comment]
> @Skillnter @ibc ... having a look on your comments (@Skillnter) I will first try to go into the code to see what could be the issue ... I thought that all of this is related to Windows running on ARM (which - In your case - is not true, right?).

Yes, that's what I wanted to convey. There is an issue but not limited to ARM only.

> 
> Your specific issue is (if I am right)
> 
> * systeminformation on regular windows non ARM (but 32bit??)

64 bit I tried. If you want, I can try on 32 bit as well.

> * `si.getDynamicData("","", cb)` and `si.getAllData("","",cb)` not returning processes

yes

> 
> So I will check first here on my side, why there is a reason not getting processes. Will come back end of next week, as I have - due to nome office regulation - currently very limited access to my windows machines.

Sure no problem. Just trying to help as much as I can.

**Comment by sebhildebrandt** [comment]
@Skillnter Thank you! OK, so this issue is a separate issue, not really related to the original title ;-) I will try to get back as soon as possible. No worries, I am sure to be able to solve it here on my side.

**Comment by Skillnter** [comment]
@sebhildebrandt thanks.

**Comment by lamuertepeluda** [comment]
I have the very same issue described in the first post since upgrading from v4.x to v5.x.

It happens with both getAllData and getDynamicData, while getStaticData works. On V4 everything is fine, so I will stick to that.

**System Configuration**
Node: 14.15.0
OS: Ubuntu 20.04 - WSL2 Linux on Window 10 Pro 19042.746

## How to Reproduce

```js
si.getAllData().then(data => console.log('Data', data))
// or
si.getDynamicData().then(data => console.log('Data', data))
```

In V5 you never get to see that data, while in V4 it takes just a few instants.

**Comment by sebhildebrandt** [comment]
@lamuertepeluda so you are running linux from WSL2 ... need to try this also. Here we have the issue that version 5 does not return any data when running `getDynamicData()` (btw. `getAllData()` also calls getDynamicData(), so here we have an issue) ... this also is not really related to the original title of the issue but anyway I will have a look on it. Thank you @lamuertepeluda for reporting!

**Comment by ibc** [comment]
> I have the very same issue described in the first post since upgrading from v4.x to v5.x.

Just a thing, I'm using `systeminformation` 4.34.5.

**Comment by lamuertepeluda** [comment]
Alright @ibc and @sebhildebrandt, if you think this is not related, maybe it's better to have a different issue for WSL 2 + v5?

**Comment by ibc** [comment]
> Alright @ibc and @sebhildebrandt, if you think this is not related, maybe it's better to have a different issue for WSL 2 + v5?

I'm afraid I've never tested v5 so that's an issue I'm not aware at all.

**Comment by sebhildebrandt** [comment]
@ibc @lamuertepeluda ... no worries. I will test it the next few days and I am sure, I will see the problem and how to fix it ;-)

Nothing to do on your side currently. Have a good day!

**Comment by sebhildebrandt** [comment]
@ibc @Skillnter @lamuertepeluda 
First of all, thank you for your patience ... I now fixed two issues (not really related to the title of this issue but mentioned here):

- `getDynamicData()` (as well as `getStaticData()`) now should return desired info also on WSL (tested with ubuntu)
- `getDynamicData()` now also returns process information on (pure) Windows

Can you test the issue related to your use case on your machines. If you still find an issue (not related to the original issue title), could you just open a new issue? Here you have the documentation for an easy testing environment (version 5):
https://systeminformation.io/tests.html

Not solved (because not yet testet): original title of this issue: Windows on ARM results.

Version 5.0.10 just released.

**Comment by ibc** [comment]
Thanks @sebhildebrandt. I'll check. "Bad news" is that I was not able to reproduce the issue anymore, but I'll notify here if I see it again.

**Comment by ibc** [comment]
I'm seeing this error in Electron 11.1.1 + Node 12.18.3 + Windows 10.0.18363:

`systeminformation.system()`, `systeminformation.cpu()`, and `systemInformation.graphics()` not resolving in 15 seconds.

**Comment by sebhildebrandt** [comment]
@ibc ... your comment here is - as far as I think - not related to win32/arm ... therefor it should be a separate issue. But nevertheless the problem you are describing is due to the fact that windows `wmic` is super slow! And as there are no other possibilities to get the right information (without writing C code, which means other dependencies and a more complex - platform dependent - install procedure), I have to go this path ...

**Comment by sebhildebrandt** [comment]
To all: please only post ARM64 (in conjunction with) Win 32 related comments here.

**Comment by ibc** [comment]
Thanks @sebhildebrandt. Just please let me know whether I should open a separate issue for win32/intel (unclear whether it already exists or you are already on it).

**Comment by ibc** [comment]
Just a question, and I assume this is true for Windows in ARM and Intel:

If we call different systeminformation async methods in parallel, may it produce some deadlock due to how wmic internally works?

**Comment by sebhildebrandt** [comment]
@ibc not this is not a problem, actually I am doing this e.g in the `graphics()` function and also in `cpu()` (and others) ... and if you try to utilize the `get()` function, all object calls will be called in parallel.

**Comment by ibc** [comment]
Thanks, good to know.

BTW I'll create a separate issue for Windows Intel.

**Comment by sebhildebrandt** [comment]
@GrumpyCrouton I still dit not manage to install Windows on my Rasperries ... maybe I am to stupid :-( 
Still this is an open issue ... can you lead me to a way how to install windows on Raspberry easily?

**Comment by OmgImAlexis** [comment]
I've opened an issue regarding the slow WMIC queries a few of you mentioned.

https://github.com/sebhildebrandt/systeminformation/issues/600

**Comment by sebhildebrandt** [comment]
@GrumpyCrouton as I managed to get results now on Windows running on ARM Platform, I assume that the original issue is solved. Can you confirm it on your side?

**Comment by GrumpyCrouton** [comment]
@sebhildebrandt Sorry, I can't. I'm no longer part of the project that I was using this module on.

**Comment by sebhildebrandt** [comment]
@GrumpyCrouton thank you for your fast reply. I will close it for now and reopen it, if it accuses again.
